feat(appium): Adjust NODE_PATH so NPM could properly resolve component peer dependencies

mykola-mokhnach opened the pull request:
## Proposed changes

According to https://nodejs.org/api/modules.html#loading-from-the-global-folders NPM should respect NODE_PATH every time it tries to resolve a module path. The idea here is that all NPM processes we start using `appium` CLI should automatically inherit parent process env, so this variable should always contain a valid path to `node_modules` root where `appium` monorepo is installed. I assume this solution would allow to avoid issues we got in https://github.com/appium/appium/pull/17297 and at the same time would allow to reduce the deployment size and the unnecessary modules duplication. This approach should also work for older NPM releases

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

mykola-mokhnach committed:
feat(appium): Adjust NODE_PATH so NPM could properly resolve component peer dependencies

boneskull commented:
AFAICT changing `NODE_PATH` at runtime has no effect.  This would only affect child processes spawned by Appium.  Is that the intent?  

I'm not really sure what this _does_.  Can we add a test to show different behavior?

mykola-mokhnach commented:
> AFAICT changing `NODE_PATH` at runtime has no effect. This would only affect child processes spawned by Appium. Is that the intent?
> 
> I'm not really sure what this _does_. Can we add a test to show different behavior?

Yes, this should affect spawned npm instances, so these could properly resolve peer deps of installed drivers and plugins

mykola-mokhnach commented:
I'm not quite sure how to properly design an e2e test to check this behavior. We need to make sure there are no extra appium node module in driver dependencies

boneskull commented:
@mykola-mokhnach 

I'd probably make two assertions here at minimum:

1. Perform an `appium driver install --source=local /path/to/fake-driver` (see function `installLocalExtension()` in `cli.e2e.spec.js`).  Then use `fs.readdir()` to look at `$APPIUM_HOME/node_modules` and assert `appium` has _not_ been installed there.
2. _In the same test_, run `appium server` and ensure it has loaded `@appium/fake-driver` properly.  I don't think an `appium driver list` will be sufficient.  It may require grepping the output of the appium child process and then killing it.

For this test, you'll need to assert `NODE_PATH` is empty first, so we can be sure it's `appium` setting the value and not the test itself.

I think what's going to happen here is 1. _might_ work, but 2. won't, because `appium` will attempt to bootstrap `fake-driver` and won't be able to, because while `appium` modifies `NODE_PATH`, `fake-driver` is not loaded by a child process of `appium`, so it doesn't matter.

Before actually writing these tests, just do it manually with a temporary `APPIUM_HOME`, e.g., `export APPIUM_HOME=/tmp/butts` and see what happens.

FWIW I looked into this myself, but didn't spend a lot of time with it.

mykola-mokhnach commented:
I've tried the scenario above manually and here are the results (npm 6 was used to install appium and the driver just to make sure no peer deps are fetched):

```
% appium driver install uiautomator2
✖ Installing 'uiautomator2' using NPM install spec 'appium-uiautomator2-driver'
Error: ✖ Encountered an error when installing package: npm command 'install --save-dev --save-exact --global-style --no-package-lock appium-uiautomator2-driver --json' failed with code 1.

STDOUT:
> appium-chromedriver@5.1.5 postinstall /Users/elf/.appium/node_modules/appium-uiautomator2-driver/node_modules/appium-chromedriver
> node install-npm.js

{
  "error": {
    "code": "ELIFECYCLE",
    "summary": "appium-chromedriver@5.1.5 postinstall: `node install-npm.js`\nExit status 1",
    "detail": "\nFailed at the appium-chromedriver@5.1.5 postinstall script.\nThis is probably not a problem with npm. There is likely additional logging output above."
  }
}

STDERR:
npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
[09:27:29] Error installing Chromedriver: Cannot find module 'appium/support'
Require stack:
- /Users/elf/.appium/node_modules/appium-uiautomator2-driver/node_modules/appium-chromedriver/build/lib/install.js
- /Users/elf/.appium/node_modules/appium-uiautomator2-driver/node_modules/appium-chromedriver/install-npm.js
[09:27:29] Error: Cannot find module 'appium/support'
Require stack:
- /Users/elf/.appium/node_modules/appium-uiautomator2-driver/node_modules/appium-chromedriver/build/lib/install.js
- /Users/elf/.appium/node_modules/appium-uiautomator2-driver/node_modules/appium-chromedriver/install-npm.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
    at Function.Module._load (internal/modules/cjs/loader.js:746:27)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:93:18)
    at Object.<anonymous> (/Users/elf/.appium/node_modules/appium-uiautomator2-driver/node_modules/appium-chromedriver/lib/install.js:1:1)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:93:18)
    at main (/Users/elf/.appium/node_modules/appium-uiautomator2-driver/node_modules/appium-chromedriver/install-npm.js:59:11)
[09:27:29] Downloading Chromedriver can be skipped by setting the'APPIUM_SKIP_CHROMEDRIVER_INSTALL' environment variable.
npm WARN notsup Unsupported engine for @appium/support@2.59.4: wanted: {"node":">=14","npm":">=8"} (current: {"node":"14.18.3","npm":"6.14.17"})
npm WARN notsup Not compatible with your version of node/npm: @appium/support@2.59.4
npm WARN appium-uiautomator2-driver@2.4.6 requires a peer of appium@^2.0.0-beta.40 but none is installed. You must install peer dependencies yourself.
npm WARN appium-android-driver@5.2.7 requires a peer of appium@^2.0.0-beta.40 but none is installed. You must install peer dependencies yourself.
npm WARN appium-chromedriver@5.1.5 requires a peer of appium@^2.0.0-beta.40 but none is installed. You must install peer dependencies yourself.
npm WARN .appium No description
npm WARN .appium No repository field.
npm WARN .appium No license field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! appium-chromedriver@5.1.5 postinstall: `node install-npm.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the appium-chromedriver@5.1.5 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/elf/.npm/_logs/2022-08-09T07_27_31_394Z-debug.log
% NODE_PATH=/usr/local/lib/node_modules appium driver install uiautomator2
✔ Installing 'uiautomator2' using NPM install spec 'appium-uiautomator2-driver'
ℹ Driver uiautomator2@2.4.6 successfully installed
- automationName: UiAutomator2
- platformNames: ["Android"]
% appium server                                                           
[Appium] npm version must be at least >=8; current is 6.14.17. Run "npm install -g npm" to upgrade.
% npm install -g npm
/usr/local/bin/npx -> /usr/local/lib/node_modules/npm/bin/npx-cli.js
/usr/local/bin/npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
+ npm@8.16.0
added 66 packages from 21 contributors, removed 299 packages and updated 138 packages in 14.758s
% appium server     
[Appium] Welcome to Appium v2.0.0-beta.42
[Appium] Attempting to load driver uiautomator2...
[debug] [Appium] Requiring driver at /Users/elf/.appium/node_modules/appium-uiautomator2-driver
[Appium] Could not load driver 'uiautomator2', so it will not be available. Error in loading the driver was: Cannot find module 'appium/driver'
[Appium] Require stack:
[Appium] - /Users/elf/.appium/node_modules/appium-uiautomator2-driver/build/lib/driver.js
[Appium] - /Users/elf/.appium/node_modules/appium-uiautomator2-driver/build/index.js
[Appium] - /usr/local/lib/node_modules/appium/build/lib/extension/extension-config.js
[Appium] - /usr/local/lib/node_modules/appium/build/lib/cli/extension-command.js
[Appium] - /usr/local/lib/node_modules/appium/build/lib/cli/driver-command.js
[Appium] - /usr/local/lib/node_modules/appium/build/lib/cli/extension.js
[Appium] - /usr/local/lib/node_modules/appium/build/lib/main.js
[Appium] - /usr/local/lib/node_modules/appium/index.js
[debug] [Appium] Error: Cannot find module 'appium/driver'
[debug] [Appium] Require stack:
[debug] [Appium] - /Users/elf/.appium/node_modules/appium-uiautomator2-driver/build/lib/driver.js
[debug] [Appium] - /Users/elf/.appium/node_modules/appium-uiautomator2-driver/build/index.js
[debug] [Appium] - /usr/local/lib/node_modules/appium/build/lib/extension/extension-config.js
[debug] [Appium] - /usr/local/lib/node_modules/appium/build/lib/cli/extension-command.js
[debug] [Appium] - /usr/local/lib/node_modules/appium/build/lib/cli/driver-command.js
[debug] [Appium] - /usr/local/lib/node_modules/appium/build/lib/cli/extension.js
[debug] [Appium] - /usr/local/lib/node_modules/appium/build/lib/main.js
[debug] [Appium] - /usr/local/lib/node_modules/appium/index.js
[debug] [Appium]     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
[debug] [Appium]     at Function.Module._load (internal/modules/cjs/loader.js:746:27)
[debug] [Appium]     at Module.require (internal/modules/cjs/loader.js:974:19)
[debug] [Appium]     at require (internal/modules/cjs/helpers.js:93:18)
[debug] [Appium]     at Object.<anonymous> (/Users/elf/.appium/node_modules/appium-uiautomator2-driver/lib/driver.js:2:1)
[debug] [Appium]     at Module._compile (internal/modules/cjs/loader.js:1085:14)
[debug] [Appium]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
[debug] [Appium]     at Module.load (internal/modules/cjs/loader.js:950:32)
[debug] [Appium]     at Function.Module._load (internal/modules/cjs/loader.js:790:12)
[debug] [Appium]     at Module.require (internal/modules/cjs/loader.js:974:19)
[debug] [Appium]     at require (internal/modules/cjs/helpers.js:93:18)
[debug] [Appium]     at Object.<anonymous> (/Users/elf/.appium/node_modules/appium-uiautomator2-driver/index.js:3:1)
[debug] [Appium]     at Module._compile (internal/modules/cjs/loader.js:1085:14)
[debug] [Appium]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
[debug] [Appium]     at Module.load (internal/modules/cjs/loader.js:950:32)
[debug] [Appium]     at Function.Module._load (internal/modules/cjs/loader.js:790:12)
[Appium] Appium REST http interface listener started on 0.0.0.0:4723
[Appium] Available drivers:
[Appium]   - uiautomator2@2.4.6 (automationName 'UiAutomator2')
[Appium] No plugins have been installed. Use the "appium plugin" command to install the one(s) you want to use.
^C[Appium] Received SIGINT - shutting down
[debug] [AppiumDriver@347b] There are no active sessions for cleanup
[HTTP] Waiting until the server is closed
[HTTP] Received server close event
% NODE_PATH=/usr/local/lib/node_modules appium server                     
[Appium] Welcome to Appium v2.0.0-beta.42
[Appium] Attempting to load driver uiautomator2...
[debug] [Appium] Requiring driver at /Users/elf/.appium/node_modules/appium-uiautomator2-driver
[Appium] Appium REST http interface listener started on 0.0.0.0:4723
[Appium] Available drivers:
[Appium]   - uiautomator2@2.4.6 (automationName 'UiAutomator2')
[Appium] No plugins have been installed. Use the "appium plugin" command to install the one(s) you want to use.
^C[Appium] Received SIGINT - shutting down
[debug] [AppiumDriver@802b] There are no active sessions for cleanup
[HTTP] Waiting until the server is closed
[HTTP] Received server close event
% 
```

mykola-mokhnach commented:
One thing I don't like is that newer NPM versions still fetch `appium` peer dependency even if it is available in `NODE_PATH`, so our previous trick with symlink creation might also be in use there to prevent that

mykola-mokhnach committed:
Add test

mykola-mokhnach committed:
Add extended error message

mykola-mokhnach committed:
fix setting of node path

mykola-mokhnach committed:
Fix tests

mykola-mokhnach committed:
fix path shortcuts processing

jlipps commented on packages/appium/test/unit/utils.spec.js:
```diff
@@ -252,4 +254,25 @@ describe('utils', function () {
+      }
+    });
+
+    it('should ajust NODE_PATH', async function () {
```
`adjust`

jlipps reviewed (approved):
this looks good to me assuming we're satisfied it works in all situations. would be nice to have some e2e tests of the kind @boneskull recommends

mykola-mokhnach committed:
Use parent folder

mykola-mokhnach committed:
Update unit test

mykola-mokhnach committed:
Tune counters

mykola-mokhnach committed:
Tune arg

mykola-mokhnach committed:
Fix condition

mykola-mokhnach committed:
Update cmdline

mykola-mokhnach committed:
Remove test

jlipps commented:
I should clarify my approval is for the code changes, but for the overall idea I think we need to make sure @boneskull is on board before merging since I don't really know the ins and outs of this problem.

boneskull commented on packages/support/lib/npm.js:
```diff
@@ -194,7 +194,7 @@ export class NPM {
 
     /**
-     * If we've found a `package.json` containined the `appiumCreated` property,
+     * If we've found a `package.json` contained the `appiumCreated` property,
```
this comment is outdated and should be removed

boneskull commented on packages/appium/lib/utils.js:
```diff
@@ -209,6 +211,50 @@ function getPackageVersion(pkgName) {
+ * for more details.
+ * @returns {Promise<void>}
+ */
+async function adjustNodePath() {
```
I mentioned this earlier, but I don't think this actually does anything.

Modifying `process.env.NODE_PATH` in process _A_ does not change the effective `NODE_PATH` of process _A_.  It _can_ affect the `NODE_PATH` of child process _B_.

If we do this in `init()`, only child processes will be effected.  Since extensions are _not_ loaded in child processes, it does _not_ change how extensions resolve `appium`.

As mentioned on the `npm` issue tracker, `npm` does not use `NODE_PATH` to determine what is to be installed.  So even if this change was intended to affect `npm`'s behavior, it won't.

boneskull replied:
We can show this does not work _in vivo_ with:

Create `foo.js`:

```js
process.env.NODE_PATH='./baz';
require('bar');
```

Create dir `baz/`. Create `baz/bar.js`:

```js
module.exports = true
```

Now, run `node foo.js`, which will fail.  Run `NODE_PATH='./baz' node foo.js`, which will succeed.

mykola-mokhnach replied:
I have found a cool article about resolving dependencies and decided to use https://gist.github.com/branneman/8048520#7-the-hack as the simplest solution for now. Although, it would probably make more sense for us to use the documented solution https://gist.github.com/branneman/8048520#8-the-wrapper

boneskull replied:
Yes, can you please use the supported solution instead?  Would rather not be touching undocumented internals...

boneskull replied:
I really want to see some sort of E2E test

mykola-mokhnach replied:
> Yes, can you please use the supported solution instead? Would rather not be touching undocumented internals...

In order to use the documented version we would have to modify imports of appium components in all depending modules. Basically instead of `import { fs, util } from 'appium/support';` it would look like 

```js
// Somewhere in Appium codebase
import { createRequire } from 'node:module';
export const requireAppium = createRequire(<path_to_appium_root>);




// Somewhere in client modules
const { fs, util } = requireAppium('./support');
```

mykola-mokhnach replied:
https://nodejs.org/api/module.html#modulecreaterequirefilename

jlipps replied:
yeah i really don't want to use a nonstandard require all over the place

boneskull commented on packages/support/lib/npm.js:
```diff
@@ -204,9 +204,10 @@ export class NPM {
-    const installOpts = (await hasAppiumDependency(cwd))
-      ? ['--save-dev']
-      : ['--save-dev', '--save-exact', '--global-style', '--no-package-lock'];
+    const installOpts = ['--save-dev', '--omit=peer'];
```
`--omit=peer` should only be present if `hasAppiumDependency()` resolves `false`.  This will cause `npm` (unless a shrinkwrap says otherwise) to not install the `appium` peer dep.  If `hasAppiumDependency()` is `true`, we want to use `npm`'s default behavior as much as possible, since it's not _our_ `package.json` which is being modified.

boneskull reviewed (commented):
I'm still not convinced this actually works; please see my comments on the `NODE_PATH`-modifying code.  

When using `--omit=peer` to install an extension, we need to verify that `appium` is not installed _anywhere_ in `$APPIUM_HOME/node_modules`, _and_ loading the extension works correctly when starting the server.  At minimum, I'd want a manual reproduction plan that I could walk through, but ideally this would be in a e2e test.

My suspicion is that if appears to work, then it's because `appium` is resolved via other means than `NODE_PATH`--and it's likely in `$APPIUM_HOME/node_modules` somewhere (though perhaps not at the root).

mykola-mokhnach committed:
Add a cool hack

mykola-mokhnach committed:
Update the logic

mykola-mokhnach committed:
Add e2e test

mykola-mokhnach committed:
Add a separate test

GoBezane commented:
Ma langue maternelle est le français et l'anglais je me débrouille..

Le jeu. 18 août 2022 à 21:55, Jonathan Lipps ***@***.***> a
écrit :

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In packages/appium/lib/utils.js
> <https://github.com/appium/appium/pull/17325#discussion_r949591958>:
>
> > @@ -209,6 +211,50 @@ function getPackageVersion(pkgName) {
>    return pkgInfo.version;
>  }
>
> +/**
> + * Adjusts NODE_PATH environment variable,
> + * so drivers and plugins could load their peer dependencies.
> + * Read https://nodejs.org/api/modules.html#loading-from-the-global-folders
> + * for more details.
> + * @returns {Promise<void>}
> + */
> +async function adjustNodePath() {
>
> yeah i really don't want to use a nonstandard require all over the place
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/appium/appium/pull/17325#discussion_r949591958>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A2RW7233GMKR64MKCNRX6ULVZ2PKZANCNFSM5527EMAQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

mykola-mokhnach committed:
Add more comments
